### PR TITLE
Canonicalize paths emitted from clang-tidy.

### PR DESCRIPTION
**Pull request by hzeller**
The paths of source files reported are not always relative to the project root, but some are prefixed with the bazel root or the current directory.

Canonicalize them so that files are always reported the same way.

Also: better quote variables in the 'shellscript' compiler invocation line; update and refine some comments.

### CODE REVIEW

File: .github/bin/run-clang-tidy-cached.cc [hunk 1]
```diff
@@ -83,6 +83,24 @@ fs::path GetCacheDir() {
...
+    if (!root.empty()) {
+      root.pop_back();  // remove newline.
+      canonicalize_expr += root + "/|";
+    }
+    canonicalize_expr += fs::current_path().string() + "/";  // $(pwd)/
+    canonicalize_expr += ")?(\\./)?";  // Some start with, or have a trailing ./
+    return std::regex{canonicalize_expr};
+  }();
+  const auto in_content = GetContent(infile);
+  std::fstream out_stream(outfile, std::ios::out);
```
Comment (fangism) [inline .github/bin/run-clang-tidy-cached.cc:100]: or just `std::ofstream`?

File: .github/bin/run-clang-tidy-cached.cc [hunk 2]
```diff
@@ -83,6 +83,24 @@ fs::path GetCacheDir() {
   return fs::path(getenv("TMPDIR") ?: "/tmp");
 }
 
+// Fix filename paths that are not emitted relative to project root.
```
Comment (fangism) [inline .github/bin/run-clang-tidy-cached.cc:86]: Perhaps include a few example transformation in comments?

**Review by fangism — APPROVED** [review]
Can you also include in a comment near the top where this script is expected to run?  e.g. linux/posix environment only.

File: .github/bin/run-clang-tidy-cached.cc (hunk 1 above)
Comment (hzeller) [reply .github/bin/run-clang-tidy-cached.cc:105]: Done.

File: .github/bin/run-clang-tidy-cached.cc (hunk 2 above)
Comment (hzeller) [reply .github/bin/run-clang-tidy-cached.cc:91]: Good idea. Added a comment at the top.

**Pull request merged**